persistent/Database/Persist/Quasi.hs

```diff
@@ -622,46 +615,32 @@ tokenize t
             let (x, y) = T.break (`elem` ['\\','(',')']) t'
              in parens count y (front . (x:))
 
--- | A line.  We don't care about spaces in the middle of the
--- line.  Also, we don't care about the amount of indentation.
-data Line' f
-    = Line
+-- | A line of parsed tokens
+data Line = Line
     { lineIndent   :: Int
-    , tokens       :: f Token
-    }
-
-deriving instance Show (f Token) => Show (Line' f)
-deriving instance Eq (f Token) => Eq (Line' f)
-
-mapLine :: (forall x. f x -> g x) -> Line' f -> Line' g
-mapLine k (Line i t) = Line i (k t)
-
-traverseLine :: Functor t => (forall x. f x -> t (g x)) -> Line' f -> t (Line' g)
-traverseLine k (Line i xs) = Line i <$> k xs
+    , tokens       :: NonEmpty Token
+    } deriving (Eq, Show)
 
-lineText :: Functor f => Line' f -> f Text
+lineText :: Line -> NonEmpty Text
 lineText = fmap tokenText . tokens
 
-type Line = Line' []
-
 lowestIndent
     :: Functor f
     => Foldable f
-    => Functor g
-    => f (Line' g)
+    => f Line
     -> Int
 lowestIndent = minimum . fmap lineIndent
```

Comment:
Can we specialize this function, too? I feel like we can get away with specializing this, possibly even to `NonEmpty` which would make it even safer (`mimumum [] = undefined`).